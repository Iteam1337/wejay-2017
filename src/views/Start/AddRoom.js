// @flow

import React, { Component } from 'react'
import { graphql } from 'react-apollo'
import { startQuery } from './Start'
import gql from 'graphql-tag'

type Props = {
  mutate: Function,
}

type State = {
  roomName: string,
}

export class AddRoom extends Component<Props, State> {
  state = {
    roomName: '',
  }

  addRoom = (event: SyntheticInputEvent<HTMLFormElement>) => {
    event.preventDefault()

    const { mutate } = this.props
    const { roomName } = this.state

    mutate({
      variables: {
        roomName,
      },
      optimisticResponse: {
        addRoom: {
          name: roomName,
          __typename: 'Room',
        },
      },
      update: (store, { data: { addRoom } }) => {
        const data = store.readQuery({ query: startQuery })

        data.rooms.push(addRoom)

        store.writeQuery({ query: startQuery, data })
      },
    })

    this.setState({
      roomName: '',
    })
  }

  updateRoomName = (event: SyntheticInputEvent<HTMLInputElement>) => {
    this.setState({
      roomName: event.target.value,
    })
  }

  render () {
    const { roomName } = this.state

    return (
      <div className="AddRoom">
        <form method="POST" onSubmit={this.addRoom}>
          <input
            className="Input"
            onChange={this.updateRoomName}
            placeholder="Room name"
            type="text"
            value={this.state.roomName}
          />
          <button
            className="Rooms__button"
            disabled={roomName.length === 0}
            type="submit">
            Add new room
          </button>
        </form>
      </div>
    )
  }
}

const AddRoomMutation = gql`
  mutation addRoom($roomName: String!) {
    addRoom(roomName: $roomName) {
      name
    }
  }
`

export default graphql(AddRoomMutation)(AddRoom)
