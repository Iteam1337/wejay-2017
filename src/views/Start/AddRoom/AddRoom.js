// @flow

import React from 'react'
import * as WejayApi from './__generated__/AddRoom'
import { Formik, Form } from 'formik'
import yup from 'yup'
import Button from '../../../components/Button/Button'
import FormInput from '../../../components/FormElements/FormInput'

type Props = {
  addRoom: (formValues: WejayApi.AddRoomVariables) => void,
}

const validateRoom = yup.object().shape({
  roomName: yup.string().required(),
})

const AddRoom = ({ addRoom }: Props) => {
  return (
    <Formik
      initialValues={{ roomName: '' }}
      onSubmit={addRoom}
      render={({ isValid }) => (
        <Form>
          <FormInput name="roomName" placeholder="Room name" />

          <Button disabled={!isValid} type="submit">
            Add new room
          </Button>
        </Form>
      )}
      validationSchema={validateRoom}
    />
  )
}

export default AddRoom
