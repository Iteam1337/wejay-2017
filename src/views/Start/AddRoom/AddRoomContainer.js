// @flow

import React from 'react'
import * as WejayApi from './__generated__/AddRoom'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import { StartQuery } from '../Start'
import AddRoom from './AddRoom'

type Props = {
  addRoom: (formValues: WejayApi.AddRoomVariables) => void,
}

const AddRoomContainer = ({ addRoom }: Props) => {
  return <AddRoom addRoom={addRoom} />
}

const AddRoomMutation = gql`
  mutation AddRoom($roomName: String!) {
    addRoom(roomName: $roomName) {
      name
    }
  }
`

const withAddRoom = graphql(AddRoomMutation, {
  props: ({ mutate }) => ({
    addRoom: async ({ roomName }: WejayApi.AddRoomVariables, actions) => {
      await mutate({
        variables: {
          roomName,
        },
        optimisticResponse: {
          addRoom: {
            name: roomName,
            __typename: 'Room',
          },
        },
        update: (store, { data: { addRoom } }) => {
          const data = store.readQuery({ query: StartQuery })

          data.rooms.push(addRoom)

          store.writeQuery({ query: StartQuery, data })
        },
      })

      actions.resetForm()
    },
  }),
})

export default withAddRoom(AddRoomContainer)
