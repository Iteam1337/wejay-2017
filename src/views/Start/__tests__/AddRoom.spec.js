import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { AddRoom } from '../AddRoom'

describe('components/AddRoom', () => {
  let component
  let instance

  beforeEach(() => {
    component = shallow(<AddRoom />)
    instance = component.instance()
  })

  it('renders AddRoom', () => {
    expect(toJson(component)).toMatchSnapshot()
  })

  describe('#updateRoomName', () => {
    it('sets room name on state', () => {
      const event = {
        target: {
          value: 'test'
        }
      }

      instance.updateRoomName(event)

      expect(component.state().roomName).toEqual('test')
    })
  })
})
