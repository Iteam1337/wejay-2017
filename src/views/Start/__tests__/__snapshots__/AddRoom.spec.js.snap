// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AddRoom renders AddRoom 1`] = `
<div
  className="AddRoom"
>
  <form
    method="POST"
    onSubmit={[Function]}
  >
    <input
      className="Input"
      onChange={[Function]}
      placeholder="Room name"
      type="text"
      value=""
    />
    <button
      className="Rooms__button"
      disabled={true}
      type="submit"
    >
      Add new room
    </button>
  </form>
</div>
`;
